File: app/src/main/java/com/kirawii/thunderswufe/work/ElectricityPredictionWorker.java
package com.kirawii.thunderswufe.work;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.kirawii.thunderswufe.ThunderApplication;
import com.kirawii.thunderswufe.notification.ElectricityNotificationManager;
import com.kirawii.thunderswufe.data.database.ElectricityRecord;
import com.kirawii.thunderswufe.ml.ElectricityPredictor;
import com.kirawii.thunderswufe.ml.PredictionResult;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;

public class ElectricityPredictionWorker extends Worker {
    private final ThunderApplication app;
    private final ElectricityNotificationManager notificationManager;

    public ElectricityPredictionWorker(
            @NonNull Context context,
            @NonNull WorkerParameters workerParams
    ) {
        super(context, workerParams);
        this.app = (ThunderApplication) context.getApplicationContext();
        this.notificationManager = new ElectricityNotificationManager(context);
    }

    @NonNull
    @Override
    public Result doWork() {
        try {
            List<ElectricityRecord> records = app.getDatabase().electricityDao().getAllRecords().blockingFirst();

            ElectricityPredictor predictor = new ElectricityPredictor(getApplicationContext());
            PredictionResult predictionResult = predictor.predictFutureUsage(records, ElectricityPredictor.ModelType.SIMPLE_LINEAR);

            if (predictionResult.getError() != null) {
                return Result.failure();
            }

            boolean notificationEnabled = app.getUserPreferencesManager().isNotificationEnabled();

            if (!notificationEnabled) {
                return Result.success();
            }

            Integer daysUntilEmpty = predictionResult.getDaysUntilEmpty();
            if (daysUntilEmpty != null && daysUntilEmpty <= 3) {
                DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM-dd");
                String emptyDate = LocalDateTime.now()
                        .plusDays(daysUntilEmpty)
                        .format(formatter);

                notificationManager.showPredictionNotification(
                        daysUntilEmpty,
                        emptyDate,
                        predictionResult.getConfidence()
                );
            }

            return Result.success();
        } catch (Exception e) {
            return Result.failure();
        }
    }
} 

File: app/src/main/java/com/kirawii/thunderswufe/work/ElectricityPredictionWorker.kt
package com.kirawii.thunderswufe.work

import android.content.Context
import androidx.work.CoroutineWorker
import androidx.work.WorkerParameters
import com.kirawii.thunderswufe.ThunderApplication
import com.kirawii.thunderswufe.notification.ElectricityNotificationManager
import com.kirawii.thunderswufe.ml.ElectricityPredictor
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

class ElectricityPredictionWorker(
    appContext: Context,
    workerParams: WorkerParameters
) : CoroutineWorker(appContext, workerParams) {

    private val app = appContext as ThunderApplication
    private val notificationManager = ElectricityNotificationManager(appContext)
    
    override suspend fun doWork(): Result = withContext(Dispatchers.IO) {
        return@withContext try {
            val records = app.database.electricityDao().getAllRecords().first()

            val predictionResult = app.electricityPredictor.predictFutureUsage(records)
            
            if (predictionResult.error != null) {
                return@withContext Result.failure()
            }

            val notificationEnabled = app.userPreferencesManager.isNotificationEnabled()
            if (!notificationEnabled) {
                return@withContext Result.success()
            }

            predictionResult.daysUntilEmpty?.let { days ->
                if (days <= 3) { // 如果预计3天内电量耗尽
                    val formatter = DateTimeFormatter.ofPattern("MM-dd")
                    val emptyDate = LocalDateTime.now().plusDays(days.toLong()).format(formatter)
                    
                    notificationManager.showPredictionNotification(
                        days,
                        emptyDate,
                        predictionResult.confidence
                    )
                }
            }

            Result.success()
        } catch (e: Exception) {
            Result.failure()
        }
    }
} 